### DISCUSSION TITLE
Moved the /system directory to another place on disk

### DISCUSSION THREAD

**Discussion by tbolon** [mastodon/mastodon#20252] [Show and tell]
I have a VM where the OS disk is small, and I don't want to use it for the /system directory, where a lot of files will be dropped.

Instead, I have a large disk mounted on `/datadrive`. Here are all the steps I have used to customize mastodon. I am completely new to nginx / sidekiq / systemd so there was certainly a simplier solution. Feel free to comment!

Initial setup:
- I have a root user and a mastodon user, as recommended
- mastodon is installed on `~/live` directory for the mastodon user
- I have a `/datadrive` mounted and created a `/datadrive/mastodon-public/` directory owned by mastodon/mastodon with chmod 775

Steps:

**Give write access to systemd**

Using your root user.

It's the main problem I have faced. Because systemd sandbox write access, the services won't have access to your datadrive. It's the reason why #20228 did not work. There is an explanation on a PR of the documentation about that: https://github.com/mastodon/documentation/pull/892/files

You have to do as suggested, and edit your `/etc/systemd/system/mastodon-web.service` and `/etc/systemd/system/mastodon-sidekiq.service` to give write permission to `/datadrive/mastodon-public/`.

In my case : `ReadWritePaths=/home/mastodon/live /datadrive/mastodon-public`

You have to restart all services and systemd after that:

```
systemctl stop mastodon-sidekiq mastodon-web
systemctl daemon-reload
systemctl start mastodon-sidekiq mastodon-web
```

**RSync your files**

You should start an initial copy of your files. Using your mastodon account.

`rsync ~/live/public/system /datadrive/mastodon-public/system/ -r`

**Update nginx config**

It's another tricky part, and the section where I am sure there are other most simple options.

Edit your nginx.conf to add the following parts (for v4):

```nginx
  location ~ ^/system/ {
    add_header Cache-Control "public, max-age=2419200, immutable";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    root /datadrive/mastodon-public/;
    try_files $uri =404;
  }
```

For v3.5 (my case), it's more complex, because I wanted to keep the custom headers for some directories:
```nginx
  # initally this location match both emojis, packs and some system directories. now it only matches non moved directories
  location ~ ^/(emoji|packs) {
    add_header Cache-Control "public, max-age=31536000, immutable";
    add_header Strict-Transport-Security "max-age=31536000" always;
    try_files $uri @proxy;
  }

  # here we have isolated the system directories to both apply http headers AND move the root
  location ~ ^/(system/accounts/avatars|system/media_attachments/files) {
    add_header Cache-Control "public, max-age=31536000, immutable";
    add_header Strict-Transport-Security "max-age=31536000" always;
    root /datadrive/mastodon-public;
    try_files $uri @proxy;
  }

  # and finally all remaining system directories are redirected to roor
  location ~ ^/(system) {
    root /datadrive/mastodon-public;
    try_files $uri @proxy;
  }
```

There are a lot of other changes in the v4 that could be applied here.

**Edit PAPERCLIP settings**

Paperclip is responsible to manage media.

```
PAPERCLIP_ROOT_PATH=/datadrive/mastodon-public/system/
PAPERCLIP_ROOT_URL=/system
```

**Reload mastodon and nginx**

```
nginx -s reload
sudo systemctl restart mastodon-sidekiq
sudo systemctl reload  mastodon-web
```

**Repeat a final rsync**

To copy the remaining files

**Rename system/ folder as system.bak/**

Check that there is no error / 404

**Erase system.bak**

**Comment by dogrileycom** [comment]
Thanks, this was helpful.  A few notes:

typo in the V4 nginx location it is missing a ;
    root /datadrive/mastodon-public/`

I used the V4 method, edited the .env.production file directly

I ran into 1 user issue and that was trying to use the same path in nginx and .env.production, nginx only need to /public and .env needs to system/ when I realized that it all worked!!

Thanks

**Reply by tbolon** [reply]
> root /datadrive/mastodon-public/`

fixed thanks.

> I ran into 1 user issue and that was trying to use the same path in nginx and .env.production, nginx only need to /public and .env needs to system/ when I realized that it all worked!!

Yes, it's tricky about the /public/system and the /system directory... I missed it a few times also.

**Comment by kpfleming** [comment]
As of the release of Mastodon 4.1.3 it is strongly suggested to add some additional headers for the `/system` path; see #25756 for details.

**Comment by allilengyi** [comment]
Thanks, this really helped fixed some issues.

**Comment by Tealk** [comment]
I made it a little easier for myself and simply mounted the disc in the /system path.

**Comment by lits101** [comment]
Could you also achieve the same result by just creating a symbolic link between ~/live/system and a folder on the /datadrive volume?

**Reply by zoltix** [reply]
Hello, 

I tried your solution to create a symbolic link between ~/live/system and a folder on /datadrive, but something strange is happening: it’s still not working and I get the same error:
```
Errno::EROFS: Read-only file system @ dir_s_mkdir - /var/lib/mastodon/public/system/cache/preview_cards/images/000/005/787
```
I’ve checked the folder permissions and even ran:
`sudo -u mastodon touch /var/lib/mastodon/public/system/testfile` which succeeds. 
Running: `sudo -u mastodon RAILS_ENV=production bin/tootctl media remove --days 0` also works. 

Yet on the new disk I continuously see the read-only filesystem error.
Do you have an idea what is happen ? 


my command to create the link
```
sudo systemctl stop mastodon-web mastodon-sidekiq mastodon-streaming
sudo mv /var/lib/mastodon/public/system /mnt/volume-hel1-1/data/mastodon
sudo ln -s /mnt/volume-hel1-1/data/mastodon/system /var/lib/mastodon/public/system
sudo systemctl start mastodon-web mastodon-sidekiq mastodon-streaming

 journalctl -fu mastodon-web

```

**Reply by tbolon** [reply]
Hey @zoltix perhaps even for symlinks, permissions need to be ajusted for systemd. Read my recomendations about "Give write access to systemd".

**Reply by zoltix** [reply]
Do you mean I need to modify the ReadWritePaths parameter to include /home/mastodon/live and /mnt/volume-hel1-1/data/mastodon?

**Reply by tbolon** [reply]
I don't know exactly, just try it and give us feedback.

**Reply by tbolon** [reply]
@zoltix did you fix your problem? I am planning on using a symlink because of `media remove` errors when redirecting the system folder, but I want to be sure you succeed in your case.

**Comment by Bautista-1** [comment]
Typed in my gmail for a verification link, refreshed and still not received an email on my gmail.
